### PR TITLE
Issue #15445: Update NPathComplexity Check to Account for when expres…

### LINKED ISSUE #15445: Update NPathComplexity Check to Account for when expression as a Possible Execution Path
check doc: https://checkstyle.org/checks/metrics/npathcomplexity.html#Description

---

```
D:\CS\test javac src/Test.java
D:\CS\test cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="NPathComplexity">
      <property name="max" value="0"/>
    </module>
  </module>
</module>
D:\CS\test cat src/Test.java
public class Test {

    // violation, NPath Complexity is 3
    void m(Object o) {
        if (o instanceof String s && !s.isEmpty()) {

        } else {

        }
    }

    // violation, NPath Complexity is 2
    void m2(Object o) {
        switch (o) {
            case String s when !s.isEmpty() -> { }
            default -> {}
        }
    }
}
D:\CS\test java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:4:5: NPath Complexity is 3 (max allowed is 0). [NPathComplexity]
[ERROR] D:\CS\test\src\Test.java:13:5: NPath Complexity is 2 (max allowed is 0). [NPathComplexity]
Audit done.
Checkstyle ends with 2 errors.
```
---

The current implementation of the NPath Complexity Check does not account for the complexity added by `when`. We should update the NPath Complexity Check to add a complexity point for `when`.

### PR DESCRIPTION
**Pull request by mahfouz72**
closes #15445.

### CODE REVIEW

**Comment by mahfouz72** [comment]
Github, generate report for NPathComplexity/all-examples-in-one

**Comment by mahfouz72** [comment]
github, generate site

**Comment by mahfouz72** [comment]
@romani fyi. the report workflow is cancelled without sending a message.

**Comment by mahfouz72** [comment]
Github, generate report for NPathComplexity/all-examples-in-one

**Comment by nrmancuso** [comment]
Report looks good, I did not manually count the changes but they seem reasonable.

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java [hunk 1]
```diff
@@ -92,6 +92,7 @@
  * <tr><td>switch ([expr]) { case : [case-range] default: [default-range] }</td>
  * <td>S(i=1:i=n)NP(case-range[i]) + NP(default-range) + NP(expr)</td></tr>
  * <tr><td>[expr1] ? [expr2] : [expr3]</td><td>NP(expr1) + NP(expr2) + NP(expr3) + 2</td></tr>
+ * <tr><td> when [expr] </td><td>NP(expr) + 1</td></tr>
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java:95]: We don't need space before `when` and after `[expr]`.

I do find this odd that it is separated from the switch, especially it not coming directly after the switch's tr. It should be moved up one line at the least.

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java (hunk 1 above)
Comment (mahfouz72) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java:95]: Moved it to be just below switch. I didn't want to keep it in the same row with switch to not make expression complicated and hard to read.

**Comment by mahfouz72** [comment]
github, generate site

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java (hunk 1 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: Yea, I understood that, but then there is nothing that says this is for a switch which is what makes it odd for me. I am ok with this until Java uses `when` in a second location.

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: @mahfouz72 Not done.
> We don't need space before when and after [expr].

**Comment by rnveach** [comment]
I am good to approve after https://github.com/checkstyle/checkstyle/pull/15522#discussion_r1725341793

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java (hunk 1 above)
Comment (mahfouz72) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/metrics/NPathComplexityCheck.java]: @rnveach done

**Review by rnveach — APPROVED** [review]

**Comment by rnveach** [comment]
@mahfouz72 CI needs to be green in this PR.

**Comment by mahfouz72** [comment]
Rebased CI should be green now.

**Comment by mahfouz72** [comment]
@rnveach ping, sempahoreci is just stopped

**Comment by romani** [comment]
Only nondex job did not run, I am ok to merge without restarting

**Pull request merged**